pommerman/network/ion_server/__init__.py

```diff
@@ -0,0 +1,130 @@
+WebSockets protocol.  
+Functions:  
+init() - If you want to run the application normally  
+run(max_players, max_matches, port, timeout, mode, ui_en=True, 
+exit_handler=True) - If you want to programatically launch the server with
+predefined parameters"""
+
+import ui
+import multiprocessing
+from . import constants
+# This is for switching between languages incase there are more
+Strings = constants.LangEN
+from . import network
+from . import match
+import time
+import random
+import signal
+import pommerman
+
+MATCHSP = []
+
+
+def _exit_handler(sp_net):
+    """Description: Return the exit handler with a reference to the sp_net
+    variable."""
+
+    def exit_handler(_s, _h):
+        """Description: Handle exiting the application."""
+        ui.info(ui.yellow, "Exiting..")
+        sp_net.terminate()
+        for i in MATCHSP:
+            i.terminate()
+        exit(0)
+
+    return exit_handler
+
+
+def init():
+    """Description: Initiate the application by asking questions."""
+    ui.info(ui.yellow, Strings.sever_starting.value)
+    port = int(ui.ask_string(Strings.server_port.value))
+    max_players = int(ui.ask_string(Strings.server_maxp.value))
+    if max_players < 4:
+        ui.fatal(ui.yellow, "Input a number greater than 4")
+    modes = []
+    for i in pommerman.configs.__dir__():
+        if i[-4:] == "_env":
+            id = getattr(pommerman.configs, i)()["env_id"]
+            if id[-2:] != "v2":
+                modes.append(id)
+    run(max_players, int(max_players / 4), port,
+        float(ui.ask_string(Strings.server_to.value)),
+        str(ui.ask_choice(Strings.server_mode.value, modes)))
+
+
+def run(max_players,
+        max_matches,
+        port,
+        timeout,
+        mode,
+        ui_en=True,
+        exit_handler=True):
+    """Arguments:  
+    * max_players: The maximum amount of concurrent players  
+    * max_matches: The maximum amount of concurrent matches (Usually this \
+should be int(max_players/4))
+    * port: The port used by the server  
+    * timeout: The time to wait before issuing the STOP action  
+    * mode: The flavor of pommerman  
+    * ui_en: If True, UI is enabled else UI is disabled  
+    * exit_handler: If True, the exit handler is set else the exit handler \
+isn't set  
+    Description: This function is responsible for running the server."""
+    netpipe, rnetpipe = multiprocessing.Pipe()
+    netqueue = multiprocessing.Queue()
+    sp_net = multiprocessing.Process(
+        target=network.thread,
+        args=(rnetpipe, netqueue, port, max_players, mode, timeout))
+    sp_net.start()
+    if exit_handler:
+        signal.signal(signal.SIGINT, _exit_handler(sp_net))
+    if ui_en:
+        ui.info(ui.yellow, Strings.server_ready.value, ui.white,
+                ui.Symbol("✔", ":)"))
+    while (True):
+        netpipe.send([constants.SubprocessCommands.get_players.value])
+        players = netpipe.recv()
```

Comment:
This appears to be an array. Please unroll it.

Reply:
I would normally. However, there can be data in-addition to the intent in the pipe so I can't unroll it.

Reply:
Let's enumerate this. What is ALL of the data that can come through this pipe?

Reply:
* [get_players] - Requesting list of players looking for a match
* [update_cc, NEW_CONCURRENTLY_LOOKING] - Update the `CONCURRENTLY_LOOKING` list to remove players who've matched.

Reply:
Okay, I misinterpreted your message. I'll unroll it.